go-controller/pkg/ovn/egressfirewall_dns.go

```diff
@@ -0,0 +1,215 @@
+		//e.dnsEntries[dnsName].dnsResolves = SortIPs(unsorted)
+		//err = e.dnsEntries[dnsName].dnsAddressSet.AddIPs(e.dnsEntries[dnsName].dnsResolves)
+		//if err != nil {
+		//	return nil, fmt.Errorf("cannot add IPs to EgressFirewalls AddressSet: %s - %v", dnsName, err)
+		//}
+		e.signalAdded(dnsName)
+	}
+	e.dnsEntries[dnsName].namespaces[namespace] = struct{}{}
+	return e.dnsEntries[dnsName].dnsAddressSet, nil
+
+}
+
+func (e *EgressDNS) Delete(namespace string) bool {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	// go through all dnsNames for namespaces
+	for dnsName, dnsEntry := range e.dnsEntries {
+		// delete the dnsEntry
+		delete(dnsEntry.namespaces, namespace)
+		if len(dnsEntry.namespaces) == 0 {
+			// the dnsEntry appears in no other namespace so delete the address_set
+			err := dnsEntry.dnsAddressSet.Destroy()
+			if err != nil {
+				klog.Errorf("Error deleteing EgressFirewall AddressSet for dnsName: %s %v", dnsName, err)
+			}
+			// the dnsEntry is no longer needed because nothing referances it delete it
+			delete(e.dnsEntries, dnsName)
+		}
+	}
+	return len(e.dnsEntries) == 0
+}
+
+func (e *EgressDNS) Update(dns string) (bool, error) {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	return e.dns.Update(dns)
+}
+
+// Sync is a goroutine that constantly syncs the dns rules. Every dns name has it's own addressset
+// that the dns query updates whenever it is run. The sync loop waits and runs on after receiving
+// one of two signals.
+// 1. a new dnsName has been added
+// 2. the minimum of either 30 minutes or the next query time of a dns name
+// if a shutdown signal is seen the goroutine exits
+func (e *EgressDNS) Sync() {
+	var duration time.Duration
+	for {
+		tm, dnsName, ok := e.GetNextQueryTime()
+		if !ok {
+			duration = 30 * time.Minute
+		} else {
+			now := time.Now()
+			if tm.After(now) {
+				duration = tm.Sub(now)
+			} else {
+				_, err := e.Update(dnsName)
+				if err != nil {
+					utilruntime.HandleError(err)
+				}
+
+				sortedIPs := SortIPs(e.dns.GetIPs(dnsName))
+				addressSet := e.dnsEntries[dnsName].dnsAddressSet
+				err = addressSet.SetIPs(sortedIPs)
+				if err != nil {
+					klog.Errorf("Cannot add IPs from EgressFirewall AddressSet: %s - %v", dnsName, err)
+				}
+				e.dnsEntries[dnsName].dnsResolves = sortedIPs
+				continue
+			}
+		}
+
+		// Wait for the given duration or till something got added
+		select {
+		case <-e.controllerStop:
+			return
+		case <-e.stopChan:
+			return
+		case newDNSName := <-e.added:
+			unsorted := e.dns.GetIPs(newDNSName)
+			e.dnsEntries[newDNSName].dnsResolves = SortIPs(unsorted)
+			err := e.dnsEntries[newDNSName].dnsAddressSet.AddIPs(e.dnsEntries[newDNSName].dnsResolves)
+			if err != nil {
+				klog.Errorf("Cannot add IPs to EgressFirewalls AddressSet: %s - %v", dnsName, err)
+			}
+
+		case <-time.After(duration):
+		}
+	}
+}
+
+func (e *EgressDNS) GetNextQueryTime() (time.Time, string, bool) {
+	e.lock.Lock()
+	defer e.lock.Unlock()
```

Comment:
@JacobTanenbaum since this function doesn't access anything internal execpt for e.dns, which never changes, and since GetNextQueryTime() locks itself, do we really need e.lock.Lock()/Unlock() here?

Reply:
Or realy, do we need this function at all? Couldn't Sync() just call e.dns.GetNextQueryTime() directly?

Reply:
Alternatively, if you want to simplify Sync(), you could do this:

```
func (e *EgressDNS) updateEntryForName(dnsName string) error {
	sortedIPs := SortIPs(e.dns.GetIPs(dnsName))
	e.dnsEntries[dnsName].dnsResolves = sortedIPs
	if err := e.dnsEntries[dnsName].dnsAddressSet.SetIPs(sortedIPs); err != nil {
		return fmt.Errorf("cannot add IPs from EgressFirewall AddressSet: %s: %v", dnsName, err)
	}
	return nil
}

// Sync is a goroutine that constantly syncs the dns rules. Every dns name has it's own addressset
// that the dns query updates whenever it is run. The sync loop waits and runs on after receiving
// one of two signals.
// 1. a new dnsName has been added
// 2. the minimum of either 30 minutes or the next query time of a dns name
// if a shutdown signal is seen the goroutine exits
func (e *EgressDNS) Sync() {
	var duration time.Duration
	for {
		duration, dnsName := e.GetNextQueryTime()
		if duration == 0 {
			if _, err := e.Update(dnsName); err != nil {
				utilruntime.HandleError(err)
			}
			if err := e.updateEntryForName(dnsName); err != nil {
				utilruntime.HandleError(err)
			}
			continue
		}

		// Wait for the given duration or till something got added
		select {
		case <-e.controllerStop:
			return
		case <-e.stopChan:
			return
		case dnsName := <-e.added:
			if err := e.updateEntryForName(dnsName); err != nil {
				utilruntime.HandleError(err)
			}
		case <-time.After(duration):
		}
	}
}

func (e *EgressDNS) GetNextQueryTime() (time.Duration, string) {
	e.lock.Lock()
	defer e.lock.Unlock()
	duration := 30 * time.Minute
	tm, dnsName, timeSet := e.dns.GetNextQueryTime()
	if !timeSet {
		// No DNS entries; wait a long time or until one shows up
		return 30 * time.Minute, ""
	}
	now := time.Now()
	if tm.After(now) {
		// Wait until the soonest entry refresh interval
		return tm.Sub(now), ""
	}

	// Refresh this name now
	return 0, dnsName
}
```

Reply:
done